src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,472 @@
+                {
+                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrTableDoesNotAcceptThisType);
+                }
+            }
+            
+            if (fValid)
+            {
+                if (context.Features.PowerFxV1CompatibilityRules && argTypes.Length == 2 && (argTypes[1].IsRecord || argTypes[1].IsPrimitive))
+                {
+                    returnType = returnType.ToRecord();
+                }
+                else
+                {
+                    returnType = returnType.ToTable();
+                }
+            }
+
+            return fValid;
+        }
+
+        public override void CheckSemantics(TexlBinding binding, TexlNode[] args, DType[] argTypes, IErrorContainer errors)
+        {
+            DType dataSourceType = argTypes[0];
+            bool isConnected = binding.EntityScope != null
+                && binding.EntityScope.TryGetDataSource(args[0], out IExternalDataSource dataSourceInfo)
+                && (dataSourceInfo.Kind == DataSourceKind.Connected || dataSourceInfo.Kind == DataSourceKind.CdsNative);
+
+            if (isConnected)
+            {
+                for (int i = 1; i < args.Length; i++)
+                {
+                    DType curType = argTypes[i];
+                    foreach (var typedName in curType.GetNames(DPath.Root))
+                    {
+                        DName name = typedName.Name;
+                        if (!dataSourceType.TryGetType(name, out DType dsNameType))
+                        {
+                            dataSourceType.ReportNonExistingName(FieldNameKind.Display, errors, name, args[i], DocumentErrorSeverity.Warning);
+                        }
+                    }
+                }
+            }
+
+            base.CheckSemantics(binding, args, argTypes, errors);
+            base.ValidateArgumentIsMutable(binding, args[0], errors);            
+
+            MutationUtils.CheckSemantics(binding, this, args, argTypes, errors);
+
+            if (binding.Features.PowerFxV1CompatibilityRules)
+            {
+                MutationUtils.CheckForReadOnlyFields(argTypes[0], args.Skip(1).ToArray(), argTypes.Skip(1).ToArray(), errors);
+            }
+        }
+
+        // This method returns true if there are special suggestions for a particular parameter of the function.
+        public override bool HasSuggestionsForParam(int argumentIndex)
+        {
+            Contracts.Assert(argumentIndex >= 0);
+
+            return argumentIndex == 0;
+        }
+
+        public override bool TryGetDataSourceNodes(PowerFx.Syntax.CallNode callNode, TexlBinding binding, out IList<FirstNameNode> dsNodes)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+
+            dsNodes = new List<FirstNameNode>();
+            if (callNode.Args.Count != 2)
+            {
+                return false;
```

Comment:
>return false; [](http://example.com/codeflow?start=16&length=13)

If this returns false, the out argument should not be initialized (i.e., dsNode = null)